// @apitable/components@1.4.0/dist/components/select/dropdown/list_dropdown.js
import { cloneElement, isValidElement, useCallback, useRef, useState } from 'react';
import classNames from 'classnames';
import { useDismiss, useRole, useClick, useInteractions, FloatingFocusManager, useId, FloatingArrow, FloatingPortal, useListNavigation, FloatingList } from '@floating-ui/react';
import React, { forwardRef, useImperativeHandle } from 'react';
import { useProviderTheme } from '../../../hooks';
import { useFloatUiDropdown } from '../../dropdown/float_ui/useFloatUiDropdown';
const CONST_INITIAL_DROPDOWN_INDEX = 1002;
export const SelectContext = React.createContext({});
export const ListDropdown = forwardRef((props, ref) => {
    var _a, _b, _c;
    const { trigger, children, onVisibleChange, options = {
        zIndex: CONST_INITIAL_DROPDOWN_INDEX
    }, className, middleware = [], clazz } = props;
    const arrowEnabled = (_a = options.arrow) !== null && _a !== void 0 ? _a : true;
    const disabled = (_b = options.disabled) !== null && _b !== void 0 ? _b : false;
    const [isOpen, setOpenValue] = useState(false);
    const setOpen = useCallback((isOpenState) => {
        if (disabled) {
            return;
        }
        setOpenValue(isOpenState);
    }, [setOpenValue, disabled]);
    const toggle = useCallback(() => {
        setOpen(!isOpen);
        onVisibleChange === null || onVisibleChange === void 0 ? void 0 : onVisibleChange(!isOpen);
    }, [isOpen, onVisibleChange, setOpen]);
    const open = useCallback(() => {
        setOpen(true);
    }, [setOpen]);
    const close = useCallback(() => {
        setOpen(false);
    }, [setOpen]);
    const [activeIndex, setActiveIndex] = React.useState(null);
    const [selectedIndex, setSelectedIndex] = React.useState((_c = options === null || options === void 0 ? void 0 : options.selectedIndex) !== null && _c !== void 0 ? _c : null);
    const resetIndex = useCallback(() => {
        var _a;
        setSelectedIndex((_a = options === null || options === void 0 ? void 0 : options.selectedIndex) !== null && _a !== void 0 ? _a : null);
    }, [setSelectedIndex, options === null || options === void 0 ? void 0 : options.selectedIndex]);
    const theme = useProviderTheme();
    const arrowRef = useRef(null);
    const triggerEl = isValidElement(trigger) ? trigger :
        trigger({
            visible: isOpen,
            toggle,
        });
    useImperativeHandle(ref, () => ({ open, toggle, close, resetIndex }));
    const elementsRef = React.useRef([]);
    const { refs, floatingStyles, context } = useFloatUiDropdown(Object.assign(Object.assign({}, options), { middleware,
        setOpen,
        isOpen,
        arrowRef }));
    const handleSelect = React.useCallback((index) => {
        setSelectedIndex(index);
        setOpen(false);
    }, [setOpen]);
    const listNav = useListNavigation(context, {
        listRef: elementsRef,
        activeIndex,
        selectedIndex,
        scrollItemIntoView: true,
        onNavigate: setActiveIndex
    });
    const click = useClick(context);
    const dismiss = useDismiss(context);
    const role = useRole(context);
    const { getReferenceProps, getFloatingProps, getItemProps } = useInteractions([
        listNav,
        click,
        dismiss,
        role
    ]);
    const selectContext = React.useMemo(() => ({
        activeIndex,
        selectedIndex,
        getItemProps,
        handleSelect
    }), [activeIndex, selectedIndex, getItemProps, handleSelect]);
    const headingId = useId();
    const setRef = (v) => {
        var _a;
        refs.setReference(v);
        (_a = props.setTriggerRef) === null || _a === void 0 ? void 0 : _a.call(props, v);
    };
    return (React.createElement(React.Fragment, null,
        // @ts-ignore
        cloneElement(triggerEl, Object.assign({ ref: setRef }, getReferenceProps())),
        isOpen && (React.createElement(FloatingPortal, null,
            React.createElement(FloatingFocusManager, { context: context },
                React.createElement("div", Object.assign({ className: classNames(className, clazz === null || clazz === void 0 ? void 0 : clazz.overlay), ref: refs.setFloating, style: floatingStyles, "aria-labelledby": headingId }, getFloatingProps()),
                    React.createElement(SelectContext.Provider, { value: selectContext },
                        React.createElement(FloatingList, { elementsRef: elementsRef }, children({ toggle }))),
                    arrowEnabled && (React.createElement(FloatingArrow, { ref: arrowRef, context: context, fill: theme.color.highestBg, strokeWidth: 1, stroke: theme.color.borderCommonDefault }))))))));
});
